sdks/java/core/src/test/java/org/apache/beam/sdk/runners/TransformHierarchyTest.java

```diff
@@ -549,33 +551,139 @@ public PCollectionTuple expand(PBegin input) {
     hierarchy.setOutput(oneAndTwo);
     hierarchy.popNode();
 
-    hierarchy.visit(new PipelineVisitor.Defaults() {
-      private final Set<Node> visitedNodes = new HashSet<>();
-      private final Set<PValue> visitedValues = new HashSet<>();
-      @Override
-      public CompositeBehavior enterCompositeTransform(Node node) {
-        for (PValue input : node.getInputs().values()) {
-          assertThat(visitedValues, hasItem(input));
-        }
-        visitedNodes.add(node);
-        return CompositeBehavior.ENTER_TRANSFORM;
-      }
-
-      @Override
-      public void visitPrimitiveTransform(Node node) {
-        assertThat(visitedNodes, hasItem(node.getEnclosingNode()));
-        for (PValue input : node.getInputs().values()) {
-          assertThat(visitedValues, hasItem(input));
-        }
-        visitedNodes.add(node);
-      }
-
-      @Override
-      public void visitValue(PValue value, Node producer) {
-        assertThat(visitedNodes, hasItem(producer));
-        assertThat(visitedValues, not(hasItem(value)));
-        visitedValues.add(value);
-      }
-    });
+    hierarchy.pushNode("second_copy_of_consumes_both", one, multiConsumer);
+    hierarchy.setOutput(done);
+    hierarchy.popNode();
+
+    hierarchy.visit(
+        new PipelineVisitor.Defaults() {
+          private final Set<Node> visitedNodes = new HashSet<>();
+          private final Set<Node> exitedNodes = new HashSet<>();
```

Comment:
should we assert that visited/exited/values all contain the right contents at the end of the test? Right now this test would pass if none of the nodes were visited (which would presumably fail a different test)